File: y15/d06/golang/main.go
package main

import (
	"flag"
	"fmt"
	"strings"
)

func main() {
	f := flag.String("f", "", "Input File Name")
	flag.Parse()
	lines := readLines(*f)
	solve(lines)
}

func solve(lines []string) {
	operateLight := func(v, mode int) int {
		if mode == 0 {
			if v == 1 {
				v = 0
			} else {
				v = 1
			}
		} else if mode == 1 {
			v = 1
		} else {
			v = 0
		}
		return v

	}
	fmt.Println("Part 1: ", operateDecoration(lines, operateLight))
	operateBrightness := func(v, mode int) int {
		if mode == 0 {
			v += 2
		} else if mode == 1 {
			v++
		} else {
			v = max(0, v-1)
		}
		return v

	}
	fmt.Println("Part 2: ", operateDecoration(lines, operateBrightness))
}

func parseInstruction(operation []string) (int, []int, []int) {
	mode, si, ei := -1, 2, 4 // mode -> operation on lights
	if len(operation) == 4 {
		mode = 0 // toggle
		si, ei = 1, 3
	} else if operation[1] == "on" {
		mode = 1 // turn on
	} else {
		mode = 2 // turn off
	}
	return mode,
		stringToIntSlice(strings.Split(operation[si], ",")),
		stringToIntSlice(strings.Split(operation[ei], ","))
}

func operateDecoration(lines []string, operate func(int, int) int) int {
	S := 1000

	grid := make([][]int, S)
	for i := 0; i < S; i++ {
		grid[i] = make([]int, S)
	}
	count := 0

	for _, line := range lines {
		operation := strings.Split(line, " ")
		mode, start, end := parseInstruction(operation)
		for r := start[0]; r <= end[0]; r++ {
			for c := start[1]; c <= end[1]; c++ {
				grid[r][c] = operate(grid[r][c], mode)
			}
		}
	}

	for r := 0; r < S; r++ {
		for c := 0; c < S; c++ {
			count += grid[r][c]
		}
	}
	return count
}

File: y15/d06/cpp/main.cpp
#include "helper.h"
#include <algorithm>
#include <iostream>
#include <ostream>
#include <vector>
using namespace std;

vector<int> parse_instruction(const vector<string> &instruction) {
  int mode = -1, si = 2, ei = 4;
  if (instruction.size() == 4) {
    si = 1, ei = 3;
    mode = 0;
  } else if (instruction[1] == "on") {
    mode = 1;
  } else {
    mode = 2;
  }

  vector<int> r1 = s_vec_to_int(split_string(instruction[si], ","));
  vector<int> r2 = s_vec_to_int(split_string(instruction[ei], ","));

  // returns mode, start_r, start_c, end_r, end_c
  return {mode, r1[0], r1[1], r2[0], r2[1]};
}

int operate_decoration(const vector<string> &lines,
                       function<int(int, int)> operate) {
  int grid[1000][1000] = {0};

  for (const auto &line : lines) {
    const vector<int> instruction = parse_instruction(split_string(line, " "));
    for (int r = instruction[1]; r <= instruction[3]; r++) {
      for (int c = instruction[2]; c <= instruction[4]; c++) {
        grid[r][c] = operate(grid[r][c], instruction[0]);
      }
    }
  }
  int count = 0;
  for (int r = 0; r < 1000; r++) {
    for (int c = 0; c < 1000; c++) {
      count += grid[r][c];
    }
  }

  return count;
}

int main(int argc, char *argv[]) {
  if (argc < 2) {
    cout << "Provide the input file name" << endl;
    return 1;
  }

  vector<string> lines = read_lines(argv[1]);

  auto operate_light = [](int v, int mode) {
    switch (mode) {
    case 0:
      return v == 1 ? 0 : 1;
    case 1:
      return 1;
    case 2:
      return 0;
    default:
      return 0;
    }
  };

  auto operate_brightness = [](int v, int mode) {
    switch (mode) {
    case 0:
      return v + 2;
    case 1:
      return v + 1;
    case 2:
      return max(0, v - 1);
    default:
      return 0;
    }
  };

  cout << "Part 1: " << operate_decoration(lines, operate_light) << endl;
  cout << "Part 2: " << operate_decoration(lines, operate_brightness) << endl;

  return 0;
}
